e2e/cluster/cluster.go

```diff
@@ -0,0 +1,503 @@
+}
+
+// Same cluster can be attempted to be created concurrently by different tests
+// sync.Once is used to ensure that only one cluster for the set of tests is created
+func ClusterKubenet(ctx context.Context) (*Cluster, error) {
+	clusterKubenetOnce.Do(func() {
+		clusterKubenet, clusterKubenetError = createNewClusterWithClient(ctx, getKubenetClusterModel(testClusterNamePrefix+"kubenet-v1"))
+	})
+	return clusterKubenet, clusterKubenetError
+}
+
+func ClusterKubenetAirgap(ctx context.Context) (*Cluster, error) {
+	clusterKubenetAirgapOnce.Do(func() {
+		cluster, err := createNewClusterWithClient(ctx, getKubenetClusterModel(testClusterNamePrefix+"kubenet-airgap"))
+		if err == nil {
+			err = addAirgapNetworkSettings(ctx, cluster)
+		}
+		clusterKubenetAirgap, clusterKubenetAirgapError = cluster, err
+	})
+	return clusterKubenetAirgap, clusterKubenetAirgapError
+}
+
+func ClusterAzureNetwork(ctx context.Context) (*Cluster, error) {
+	clusterAzureNetworkOnce.Do(func() {
+		clusterAzureNetwork, clusterAzureNetworkError = createNewClusterWithClient(ctx, getAzureNetworkClusterModel(testClusterNamePrefix+"azure-network"))
+	})
+	return clusterAzureNetwork, clusterAzureNetworkError
+}
+
+func createNewClusterWithClient(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*Cluster, error) {
+	createdCluster, err := createNewClusterWithRetry(ctx, cluster)
+	if err != nil {
+		return nil, err
+	}
+
+	kube, err := getClusterKubeClient(ctx, config.ResourceGroupName, *cluster.Name)
+	if err != nil {
+		return nil, fmt.Errorf("get kube client using cluster %q: %w", *cluster.Name, err)
+	}
+
+	if err := ensureDebugDaemonset(ctx, kube); err != nil {
+		return nil, fmt.Errorf("unable to ensure debug damonset of viable cluster %q: %w", *cluster.Name, err)
+	}
+
+	return &Cluster{Model: createdCluster, Kube: kube}, nil
+
+}
+
+func createNewCluster(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*armcontainerservice.ManagedCluster, error) {
+	log.Printf("Creating new cluster %s in rg %s\n", *cluster.Name, *cluster.Location)
+	pollerResp, err := config.Azure.AKS.BeginCreateOrUpdate(
+		ctx,
+		config.ResourceGroupName,
+		*cluster.Name,
+		*cluster,
+		nil,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("failed to begin aks cluster creation: %w", err)
+	}
+
+	clusterResp, err := pollerResp.PollUntilDone(ctx, nil)
+	if err != nil {
+		return nil, fmt.Errorf("failed to wait for aks cluster creation %w", err)
+	}
+
+	return &clusterResp.ManagedCluster, nil
+}
+
+// createNewClusterWithRetry is a wrapper around createNewCluster
+// that retries creating a cluster if it fails with a 409 Conflict error
+// clusters are reused, and sometimes a cluster can be in UPDATING or DELETING state
+// simple retry should be sufficient to avoid such conflicts
+func createNewClusterWithRetry(ctx context.Context, cluster *armcontainerservice.ManagedCluster) (*armcontainerservice.ManagedCluster, error) {
```

Comment:
can we turn this into a poller and add to `pollers.go` instead?

Reply:
Currently it's a specialized method for cluster creation and not the complicated.

Generalizing it adds extra complexity. And the usage of overly generalized things can be more complicated.

Maybe there is a good well-known way to handle it or a library. But I don't see how I can really simplify it.